src/ora/src/Ora.cc

```diff
@@ -0,0 +1,276 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+#include "ora/Ora.hh"
+
+#include <curl/curl.h>
+
+#include <boost/json.hpp>
+#include <fstream>
+#include <sstream>
+#include <string>
+
+#include "sta/StaMain.hh"
+#include "utl/Logger.h"
+
+namespace sta {
+// Tcl files encoded into strings.
+extern const char* ora_tcl_inits[];
+}  // namespace sta
+
+namespace ora {
+
+using utl::Logger;
+
+extern "C" {
+extern int Ora_Init(Tcl_Interp* interp);
+}
+
+void setLogger(Logger* logger);
+
+Logger* logger_ = nullptr;
+
+Ora::Ora() : sourceFlag_(true)
+{
+}
+
+Ora::~Ora()
+{
+}
+
+size_t WriteCallback(void* contents, size_t size, size_t nmemb, void* userp)
+{
+  ((std::string*) userp)->append((char*) contents, size * nmemb);
+  return size * nmemb;
+}
+
+bool checkConsent()
+{
+  std::string folderPath
+      = std::string(getenv("HOME")) + "/.local/share/openroad";
+  std::ifstream folderFile(folderPath);
+  if (!folderFile) {
+    return false;
+  }
+
+  std::string consentFilePath = folderPath + "/orassistant_consent.txt";
+  std::ifstream consentFile(consentFilePath);
+
+  if (consentFile.is_open()) {
+    std::string consent;
+    std::getline(consentFile, consent);
+    consentFile.close();
+    return consent == "y";
+  }
+
+  return false;
+}
+
+std::string sendPostRequest(const std::string& url, const std::string& jsonData)
+{
+  CURL* curl;
+  CURLcode res;
+  std::string response;
+
+  curl_global_init(CURL_GLOBAL_ALL);
+  curl = curl_easy_init();
+
+  if (curl) {
+    struct curl_slist* headers = nullptr;
+    headers = curl_slist_append(headers, "Content-Type: application/json");
+
+    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, jsonData.c_str());
+    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
+
+    res = curl_easy_perform(curl);
+
+    if (res != CURLE_OK) {
+      response = "Error: " + std::string(curl_easy_strerror(res));
+    }
+
+    curl_slist_free_all(headers);
+    curl_easy_cleanup(curl);
+  } else {
+    response = "Error: curl_easy_init() failed";
+  }
+
+  curl_global_cleanup();
+  return response;
+}
+
+void Ora::init(Tcl_Interp* tcl_interp, odb::dbDatabase* db, utl::Logger* logger)
+{
+  db_ = db;
+  logger_ = logger;
+
+  Ora_Init(tcl_interp);
+  sta::evalTclInit(tcl_interp, sta::ora_tcl_inits);
+
+  hostUrl
+      = "https://bursting-stallion-friendly.ngrok-free.app/graphs/"
+        "agent-retriever";
+  localDirPath = std::string(getenv("HOME")) + "/.local/share/openroad";
```

Comment:
why not make a helper function to generate this path since it's used in multiple locations